"SystemError: tile cannot extend outside image" on s390x

Reported downstream at https://bugzilla.redhat.com/show_bug.cgi?id=1706450:

*Description of problem:*
I am trying to update a few packages (python-matplotlib [1] and python-pikepdf [2]), but they are failing on s390x with "SystemError: tile cannot extend outside image". I have not attempted to update matplotlib on F30, but at least pikepdf continues to work [3] on F30 with its version of pillow.


*Version-Release number of selected component (if applicable):*
python-pillow-6.0.0-1.fc31


*Steps to Reproduce:*
1. Build python-matplotlib or python-pikepdf on Fedora Rawhide


*Actual results:*

    /usr/lib64/python3.7/site-packages/PIL/Image.py:2007: in save
        save_handler(self, fp, filename)
    /usr/lib64/python3.7/site-packages/PIL/PngImagePlugin.py:896: in _save
        [("zip", (0, 0)+im.size, 0, rawmode)])
    _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    im = <PIL.Image.Image image mode=RGBA size=800x600 at 0x3FF8EA5C2B0>
    fp = <PIL.PngImagePlugin._idat object at 0x3ff8eade160>
    tile = [('zip', (0, 0, 800, 600), 0, 'RGBA')], bufsize = 65536
        def _save(im, fp, tile, bufsize=0):
            """Helper to save image based on tile list
        
            :param im: Image object.
            :param fp: File object.
            :param tile: Tile list.
            :param bufsize: Optional buffer size
            """
        
            im.load()
            if not hasattr(im, "encoderconfig"):
                im.encoderconfig = ()
            tile.sort(key=_tilesort)
            # FIXME: make MAXBLOCK a configuration parameter
            # It would be great if we could have the encoder specify what it needs
            # But, it would need at least the image size in most cases. RawEncode is
            # a tricky case.
            bufsize = max(MAXBLOCK, bufsize, im.size[0] * 4)  # see RawEncode.c
            if fp == sys.stdout:
                fp.flush()
                return
            try:
                fh = fp.fileno()
                fp.flush()
            except (AttributeError, io.UnsupportedOperation):
                # compress to Python file-compatible object
                for e, b, o, a in tile:
                    e = Image._getencoder(im.mode, e, a, im.encoderconfig)
                    if o > 0:
                        fp.seek(o)
    >               e.setimage(im.im, b)
    E               SystemError: tile cannot extend outside image
    /usr/lib64/python3.7/site-packages/PIL/ImageFile.py:496: SystemError

--- a/src/_imagingft.c
+++ b/src/_imagingft.c
@@ -257,7 +257,7 @@ getfont(PyObject* self_, PyObject* args, PyObject* kw)
         return NULL;
     }
 
-    if (!PyArg_ParseTupleAndKeywords(args, kw, "eti|is"PY_ARG_BYTES_LENGTH"i",
+    if (!PyArg_ParseTupleAndKeywords(args, kw, "etn|ns"PY_ARG_BYTES_LENGTH"n",
                                      kwlist,
                                      Py_FileSystemDefaultEncoding, &filename,
                                      &size, &index, &encoding, &font_bytes,

--- a/src/encode.c
+++ b/src/encode.c
@@ -126,7 +126,7 @@ _encode(ImagingEncoderObject* encoder, PyObject* args)
 
     Py_ssize_t bufsize = 16384;
 
-    if (!PyArg_ParseTuple(args, "|i", &bufsize))
+    if (!PyArg_ParseTuple(args, "|n", &bufsize))
         return NULL;
 
     buf = PyBytes_FromStringAndSize(NULL, bufsize);
@@ -180,7 +180,7 @@ _encode_to_file(ImagingEncoderObject* encoder, PyObject* args)
     Py_ssize_t fh;
     Py_ssize_t bufsize = 16384;
 
-    if (!PyArg_ParseTuple(args, "i|i", &fh, &bufsize))
+    if (!PyArg_ParseTuple(args, "n|n", &fh, &bufsize))
         return NULL;
 
     /* Allocate an encoder buffer */
@@ -229,7 +229,7 @@ _setimage(ImagingEncoderObject* encoder, PyObject* args)
     x0 = y0 = x1 = y1 = 0;
 
     /* FIXME: should publish the ImagingType descriptor */
-    if (!PyArg_ParseTuple(args, "O|(iiii)", &op, &x0, &y0, &x1, &y1))
+    if (!PyArg_ParseTuple(args, "O|(nnnn)", &op, &x0, &y0, &x1, &y1))
         return NULL;
     im = PyImaging_AsImaging(op);
     if (!im)
@@ -409,7 +409,7 @@ PyImaging_GifEncoderNew(PyObject* self, PyObject* args)
     char *rawmode;
     Py_ssize_t bits = 8;
     Py_ssize_t interlace = 0;
-    if (!PyArg_ParseTuple(args, "ss|ii", &mode, &rawmode, &bits, &interlace))
+    if (!PyArg_ParseTuple(args, "ss|nn", &mode, &rawmode, &bits, &interlace))
         return NULL;
 
     encoder = PyImaging_EncoderNew(sizeof(GIFENCODERSTATE));
@@ -441,7 +441,7 @@ PyImaging_PcxEncoderNew(PyObject* self, PyObject* args)
     char *rawmode;
     Py_ssize_t bits = 8;
 
-    if (!PyArg_ParseTuple(args, "ss|ii", &mode, &rawmode, &bits)) {
+    if (!PyArg_ParseTuple(args, "ss|n", &mode, &rawmode, &bits)) {
         return NULL;
     }
 
@@ -474,7 +474,7 @@ PyImaging_RawEncoderNew(PyObject* self, PyObject* args)
     Py_ssize_t stride = 0;
     Py_ssize_t ystep = 1;
 
-    if (!PyArg_ParseTuple(args, "ss|ii", &mode, &rawmode, &stride, &ystep))
+    if (!PyArg_ParseTuple(args, "ss|nn", &mode, &rawmode, &stride, &ystep))
         return NULL;
 
     encoder = PyImaging_EncoderNew(0);
@@ -506,7 +506,7 @@ PyImaging_TgaRleEncoderNew(PyObject* self, PyObject* args)
     char *rawmode;
     Py_ssize_t ystep = 1;
 
-    if (!PyArg_ParseTuple(args, "ss|i", &mode, &rawmode, &ystep))
+    if (!PyArg_ParseTuple(args, "ss|n", &mode, &rawmode, &ystep))
         return NULL;
 
     encoder = PyImaging_EncoderNew(0);
@@ -567,7 +567,7 @@ PyImaging_ZipEncoderNew(PyObject* self, PyObject* args)
     Py_ssize_t compress_type = -1;
     char* dictionary = NULL;
     Py_ssize_t dictionary_size = 0;
-    if (!PyArg_ParseTuple(args, "ss|iii"PY_ARG_BYTES_LENGTH, &mode, &rawmode,
+    if (!PyArg_ParseTuple(args, "ss|nnn"PY_ARG_BYTES_LENGTH, &mode, &rawmode,
                           &optimize,
                           &compress_level, &compress_type,
                           &dictionary, &dictionary_size))
@@ -717,7 +717,7 @@ PyImaging_JpegEncoderNew(PyObject* self, PyObject* args)
     char* rawExif = NULL;
     Py_ssize_t rawExifLen = 0;
 
-    if (!PyArg_ParseTuple(args, "ss|iiiiiiiiO"PY_ARG_BYTES_LENGTH""PY_ARG_BYTES_LENGTH,
+    if (!PyArg_ParseTuple(args, "ss|nnnnnnnnO"PY_ARG_BYTES_LENGTH""PY_ARG_BYTES_LENGTH,
                           &mode, &rawmode, &quality,
                           &progressive, &smooth, &optimize, &streamtype,
                           &xdpi, &ydpi, &subsampling, &qtables, &extra, &extra_size,
@@ -823,7 +823,7 @@ PyImaging_LibTiffEncoderNew(PyObject* self, PyObject* args)
     PyObject *keys, *values;
 
 
-    if (! PyArg_ParseTuple(args, "sssisO", &mode, &rawmode, &compname, &fp, &filename, &dir)) {
+    if (! PyArg_ParseTuple(args, "sssnsO", &mode, &rawmode, &compname, &fp, &filename, &dir)) {
         return NULL;
     }
 
@@ -995,7 +995,7 @@ PyImaging_Jpeg2KEncoderNew(PyObject *self, PyObject *args)
     OPJ_CINEMA_MODE cine_mode;
     Py_ssize_t fd = -1;
 
-    if (!PyArg_ParseTuple(args, "ss|OOOsOIOOOssi", &mode, &format,
+    if (!PyArg_ParseTuple(args, "ss|OOOsOnOOOssn", &mode, &format,
                           &offset, &tile_offset, &tile_size,
                           &quality_mode, &quality_layers, &num_resolutions,
                           &cblk_size, &precinct_size,

